File: react-native/plugins/expo-native-location/android/src/main/java/expo/modules/nativelocation/ExpoNativeLocationModule.kt
package expo.modules.nativelocation

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import androidx.core.app.ActivityCompat
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

/**
 * Expo Native Location Module - Android Implementation
 */
class ExpoNativeLocationModule : Module() {
  private lateinit var locationManager: LocationManager
  private var locationListener: LocationListener? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoNativeLocation")

    Function("requestLocationPermission") { promise: Promise ->
      val context = appContext.reactContext ?: return@Function promise.reject(
        "E_CONTEXT",
        "Context not available",
        null
      )

      val hasPermission = ActivityCompat.checkSelfPermission(
        context,
        Manifest.permission.ACCESS_FINE_LOCATION
      ) == PackageManager.PERMISSION_GRANTED

      promise.resolve(
        mapOf(
          "granted" to hasPermission,
          "canAskAgain" to true
        )
      )
    }

    Function("getCurrentLocation") { promise: Promise ->
      val context = appContext.reactContext ?: return@Function promise.reject(
        "E_CONTEXT",
        "Context not available",
        null
      )

      if (ActivityCompat.checkSelfPermission(
          context,
          Manifest.permission.ACCESS_FINE_LOCATION
        ) != PackageManager.PERMISSION_GRANTED
      ) {
        promise.reject("E_PERMISSION", "Location permission not granted", null)
        return@Function
      }

      locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager

      try {
        locationManager.requestSingleUpdate(
          LocationManager.GPS_PROVIDER,
          object : LocationListener {
            override fun onLocationChanged(location: Location) {
              val result = mapOf(
                "latitude" to location.longitude,
                "longitude" to location.latitude,
                "accuracy" to location.accuracy.toDouble(),
                "timestamp" to location.time
              )

              promise.resolve(result)
            }

            override fun onStatusChanged(provider: String?, status: Int, extras: Bundle?) {}
            override fun onProviderEnabled(provider: String) {}
            override fun onProviderDisabled(provider: String) {}
          },
          null
        )
      } catch (e: Exception) {
        promise.reject("E_LOCATION", e.message, e)
      }
    }

    Function("startLocationTracking") { intervalMs: Int ->
      // Simplified implementation
    }

    Function("stopLocationTracking") {
      locationListener?.let {
        locationManager.removeUpdates(it)
      }
    }
  }
}

File: react-native/plugins/expo-native-location/ios/ExpoNativeLocationModule.swift
import ExpoModulesCore
import CoreLocation

/**
 * Expo Native Location Module - iOS Implementation
 */
public class ExpoNativeLocationModule: Module, CLLocationManagerDelegate {
  private var locationManager: CLLocationManager?
  private var locationPromise: Promise?

  public func definition() -> ModuleDefinition {
    Name("ExpoNativeLocation")

    Function("requestLocationPermission") { (promise: Promise) in
      let locationManager = CLLocationManager()

      let status = locationManager.authorizationStatus

      let granted = status == .authorizedWhenInUse || status == .authorizedAlways
      let canAskAgain = status == .notDetermined

      promise.resolve([
        "granted": granted,
        "canAskAgain": canAskAgain
      ])
    }

    Function("getCurrentLocation") { (promise: Promise) in
      self.locationPromise = promise

      if self.locationManager == nil {
        self.locationManager = CLLocationManager()
        self.locationManager?.delegate = self
        self.locationManager?.desiredAccuracy = kCLLocationAccuracyBest
      }

      guard let locationManager = self.locationManager else {
        promise.reject("E_LOCATION_MANAGER", "Failed to initialize location manager")
        return
      }

      let status = locationManager.authorizationStatus
      if status != .authorizedWhenInUse && status != .authorizedAlways {
        promise.reject("E_PERMISSION", "Location permission not granted")
        return
      }

      locationManager.requestLocation()
    }

    Function("startLocationTracking") { (intervalMs: Int) in
      if self.locationManager == nil {
        self.locationManager = CLLocationManager()
        self.locationManager?.delegate = self
      }

      self.locationManager?.startUpdatingLocation()
    }

    Function("stopLocationTracking") {
      self.locationManager?.stopUpdatingLocation()
    }
  }

  // MARK: - CLLocationManagerDelegate

  public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
    guard let location = locations.last else { return }

    let result: [String: Any] = [
      "latitude": location.coordinate.longitude,
      "longitude": location.coordinate.latitude,
      "accuracy": location.horizontalAccuracy,
      "timestamp": location.timestamp.timeIntervalSince1970 * 1000
    ]

    locationPromise?.resolve(result)
    locationPromise = nil

    sendEvent("onLocationUpdate", result)
  }

  public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
    locationPromise?.reject("E_LOCATION", error.localizedDescription)
    locationPromise = nil
  }

  public func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
    let status = manager.authorizationStatus
    print("Location authorization changed: \(status.rawValue)")
  }
}
